### ISSUE TITLE
rye add doesn't find correct jax (CUDA) package

### ISSUE THREAD

**Issue by rachtsingh** [mitsuhiko/rye#755]
### Steps to Reproduce

I know this is very close to a duplicate of https://github.com/mitsuhiko/rye/issues/420, but I think the issue is slightly different (`rye add` vs `rye sync`).

Run `rye init rye-test`, cd into the directory, and add the following to `pyproject.toml`:

```
[[tool.rye.sources]]
name = "google-jax"
url = "https://storage.googleapis.com/jax-releases/jax_cuda_releases.html"
type = "find-links"
```

Run `rye add "jax[cuda12_pip]"`.

### Expected Result

The final result should look like this:

```
Reusing already existing virtualenv
Generating production lockfile: /home/singhrac/proj/annotated-s4/requirements.lock
   Built file:///home/singhrac/proj/annotated-s4                                                                                                               Built 1 editable in 187ms
Resolved 33 packages in 559ms
Generating dev lockfile: /home/singhrac/proj/annotated-s4/requirements-dev.lock
   Built file:///home/singhrac/proj/annotated-s4                                                                                                               Built 1 editable in 179ms
Resolved 58 packages in 202ms
Installing dependencies
Resolved 10 packages in 3ms
Downloaded 10 packages in 51.85s
Installed 16 packages in 40ms
 + jax==0.4.24
 + jaxlib==0.4.24+cuda12.cudnn89
 + ml-dtypes==0.3.2
 + nvidia-cublas-cu12==12.3.4.1
 + nvidia-cuda-cupti-cu12==12.3.101
 + nvidia-cuda-nvcc-cu12==12.3.107
 + nvidia-cuda-nvrtc-cu12==12.3.107
 + nvidia-cuda-runtime-cu12==12.3.101
 + nvidia-cudnn-cu12==8.9.7.29
 + nvidia-cufft-cu12==11.0.12.1
 + nvidia-cusolver-cu12==11.5.4.101
 + nvidia-cusparse-cu12==12.2.0.103
 + nvidia-nccl-cu12==2.19.3
 + nvidia-nvjitlink-cu12==12.3.101
 + opt-einsum==3.3.0
 + scipy==1.12.0
Done!
```

### Actual Result

```
Initializing new virtualenv in /home/singhrac/proj/rye-test/.venv
Python version: cpython@3.12.1
Added jax>=0.4.24 as regular dependency
Reusing already existing virtualenv
Generating production lockfile: /home/singhrac/proj/rye-test/requirements.lock
warning: Requirements file /tmp/.tmpNrItY2/requirements.txt does not contain any dependencies
   Built file:///home/singhrac/proj/rye-test
Resolved 6 packages in 192ms
Generating dev lockfile: /home/singhrac/proj/rye-test/requirements-dev.lock
warning: Requirements file /tmp/.tmpkoWACb/requirements.txt does not contain any dependencies
   Built file:///home/singhrac/proj/rye-test
Resolved 6 packages in 184ms
Installing dependencies
   Built file:///home/singhrac/proj/rye-test
Installed 6 packages in 17ms
 + jax==0.4.24
 + ml-dtypes==0.3.2
 + numpy==1.26.4
 + opt-einsum==3.3.0
 + rye-test==0.1.0 (from file:///home/singhrac/proj/rye-test)
 + scipy==1.12.0
Done!
```

This means the CPU version of jax is installed, but not the CUDA version.

However, `rye sync` does correctly install the CUDA version if we manually add `"jax[cuda12_pip]>=0.4.23"` to `project.dependencies`. 

Also, in a fresh environment, the following does work:

```
uv pip install "jax[cuda12_pip]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
```

So it seems like it's isolated to `rye add`?

### Version Info

```
rye 0.26.0
commit: 0.26.0 (d245f625e 2024-02-23)
platform: linux (x86_64)
self-python: cpython@3.12
symlink support: true
uv enabled: true
```

**Comment by rachtsingh** [comment]
Probably a duplicate of https://github.com/mitsuhiko/rye/issues/745, closing for that (will reopen if still not working after that's fixed).

**Comment by rachtsingh** [comment]
I just tried this again, and while the new rye gets past the previous issue, it now fails with this:

```
~/proj/rye-test main ?5 ❯ rye add "jax[cuda12_pip]"
Initializing new virtualenv in /home/singhrac/proj/rye-test/.venv
Python version: cpython@3.12.1
Added jax[cuda12_pip]>=0.4.25 as regular dependency
Reusing already existing virtualenv
Generating production lockfile: /home/singhrac/proj/rye-test/requirements.lock
  × No solution found when resolving dependencies:
  ╰─▶ Because only nvidia-cublas-cu12<12.3.4.1 is available and jax[cuda12-pip]==0.4.25 depends on nvidia-cublas-cu12>=12.3.4.1, we can conclude that
      jax[cuda12-pip]==0.4.25 cannot be used.
      And because only jax[cuda12-pip]<=0.4.25 is available and you require jax[cuda12-pip]>=0.4.25, we can conclude that the requirements are unsatisfiable.
error: could not write production lockfile for project

Caused by:
    failed to generate lockfile
```

which I think is the same as https://github.com/astral-sh/rye/issues/420 .

I think this is probably an issue with either Jax's dependencies or NVIDIA's, but not sure which.
